### ISSUE TITLE
[Research] Summer Workflow Proposals — PEbA & Uni-Mol

### ISSUE THREAD

**Issue by RongjunG** [chiral-data/collab-workflows#172]
Hi @allisongcheng , great research on PEbA and Uni-Mol! Before deciding on a node structure, here's some background on how these tools are actually used in the literature, which existing workflows they can integrate with, and a few potential directions to consider.

---

### PEbA — Protein Embedding-Based Alignment
- **GitHub:** https://github.com/mgtools/PEbA
- **Paper:** Iovino & Ye 2024, *BMC Bioinformatics* ([10.1186/s12859-024-05699-5](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-024-05699-5))
- **Install:** `pip install -r requirements.txt` (PyTorch + ProtT5/ESM2 from HuggingFace)
- **What it does:** Pairwise protein sequence alignment using embedding cosine similarity (ProtT5, ESM2, or DEDAL) instead of substitution matrices. Targets the "twilight zone" (<20–35% identity) where BLAST/BLOSUM fail.
- **I/O:** FASTA in → MSF or FASTA alignment out
- **Key limitation:** Pairwise only — not a database search tool. Scaling to DB search requires pre-computed embeddings.

### Uni-Mol — Universal 3D Molecular Representation Framework
- **GitHub:** https://github.com/deepmodeling/Uni-Mol (1.1K stars, MIT license)
- **Core paper:** Zhou et al. 2022 (ICLR 2023) — [openreview.net/forum?id=6K2RM6wVqKu](https://openreview.net/forum?id=6K2RM6wVqKu)

| Component | What it does | Install | I/O |
|---|---|---|---|
| **Uni-Mol Docking V2** | Protein-ligand docking, 77.6% RMSD<2Å on PoseBusters (Alcaide et al. 2024, [arXiv:2405.11769](https://arxiv.org/abs/2405.11769)) | Requires Uni-Core (custom CUDA kernels) | PDB + SDF + grid JSON → SDF poses |
| **Uni-Mol Tools** | Easy-use property prediction wrapper (Gao et al. 2023) | `pip install unimol_tools` | SMILES list/CSV → predictions DataFrame |
| **Uni-Mol Repr** | 3D molecular embedding extraction | `pip install unimol_tools` | SMILES list → cls_repr vectors + atomic_reprs |

---

## How Uni-Mol Is Actually Used (Literature Survey)

Based on 2024–2026 papers, Uni-Mol appears in three usage modes:

### Mode 1: As a Docking Engine (most common)
Standalone replacement for Vina/GNINA. Examples:
- EasyDock 1.3 (2026, *JCIM*) — automated docking pipeline built on Uni-Mol encoder
- Li et al. 2026 (*Accounts of Chemical Research*) — covalent inhibitor binding site analysis

### Mode 2: As a Pre-screener → Traditional Docking Validation
Fine-tune Uni-Mol for activity prediction, then validate top hits with Vina. Examples:
- Yang et al. 2026 (*Acta Pharmacologica Sinica*) — fine-tuned Uni-Mol → predicted T cell inhibitors → AutoDock Vina validation
- Ma et al. 2026 (*J Comput-Aided Mol Des*) — Uni-Mol activity prediction + traditional docking → LRRK2 inhibitors
- **Note:** This mode requires target-specific training data for fine-tuning.

### Mode 3: As a Molecular Encoder / Feature Extractor
Use Uni-Mol 3D representations as features for downstream ML:
- DrugCLIP / GalaxyVS (2026) — Uni-Mol encoder for contrastive learning, screens 100 billion compounds
- Activity Cliff Aware Active Learning (2026) — Uni-Mol repr for hit discovery acceleration
- MolBench (2023) — Uni-Mol ranked #1 in property prediction benchmark, outperformed D-MPNN
- Sheshanarayana & You 2025 — "Uni-Mol outperformed 2D baselines on 14 of 15 property prediction tasks"

---

## How PEbA Fits In

PEbA itself is relatively new (27 citations), but the broader approach — PLM embeddings for remote homology — is well-validated:
- PLMSearch (Liu et al. 2024, *Nature Communications*) — embedding-based pairwise alignment for remote homology
- Johnson et al. 2024 (*eLife*) — local alignment of positional embeddings for sensitive homology search
- CATHe2 — PEbA-style alignment for protein superfamily detection

---

### Consensus Docking + ADMET (established)
- DockM8 (2024): "All-in-One Consensus Virtual Screening" platform with Vina + GNINA
- ESSENCE-Dock (2024): Vina + GNINA consensus scoring
- Nelen 2025: consensus docking pipeline evaluating "activity, selectivity and ADMET properties"
- 2026 arXiv: "when carefully implemented, consensus docking improved screening outcomes"

### AI-Predicted Structure → Docking (growing)
- Multiple 2025–2026 papers use AlphaFold/Boltz-predicted structures as docking targets
- Key caveat: "binding site quality is more important than global fold accuracy"
- pLDDT per-residue confidence can flag unreliable pocket regions

---

## Existing Workflows & Tools to Consider

| Existing tool/workflow | What it provides | Could integrate with |
|---|---|---|
| GNINA (app) | CNN-based molecular docking | Uni-Mol Docking for consensus scoring |
| ADMET-AI (workflow-014) | 49 ADMET properties from SMILES | Any docking workflow for integrated drug-likeness filtering |
| Boltz-2 (workflow-012) | AI structure prediction → PDB | Uni-Mol Docking when no crystal structure exists |
| P2Rank (used in workflow-004) | Binding pocket detection | Any docking workflow for automated pocket identification |
| QSAR pipeline (workflow-005/011) | ML property prediction from RDKit/MACCS features | Uni-Mol Repr as upgraded feature extractor |
| BLAST (app) | Traditional sequence similarity search | PEbA for side-by-side comparison in twilight zone |
| mDeepFRI (workflow-015) | GO term function prediction from FASTA | PEbA for orthogonal homology detection |
| FEGrow active learning (workflow-009) | Surrogate-guided lead optimization | Uni-Mol Repr as surrogate model features |

---

## Potential Directions

These are starting points — pick one or combine, then propose a node structure.

### A. Consensus Docking + ADMET
Uni-Mol Docking V2 + GNINA + ADMET-AI. Extends the workflow-025 pattern (Vina vs GNINA) with a third scoring paradigm + integrated ADMET filtering. I/O chains cleanly: PDB + SDF shared across docking engines, SMILES parallel to ADMET. Main challenge: Uni-Core containerization.

### B. Structure Prediction → Docking (No Crystal Structure Needed)
Boltz-2 → P2Rank → Uni-Mol Docking V2. Full chain from sequence to docked poses. Boltz-2 pLDDT can auto-flag low-confidence pockets. I/O: YAML → PDB → pocket coords → grid JSON → SDF. All formats compatible.

### C. Uni-Mol Repr Upgrade for QSAR
Replace RDKit/MACCS features in workflow-005/011 with Uni-Mol 3D representations. Benchmark-validated improvement (14/15 tasks). `pip install unimol_tools` — lowest barrier. More of an upgrade than a new use case.

### D. PEbA + BLAST + mDeepFRI (Twilight Zone Function Discovery)
BLAST as baseline, PEbA for improved alignment, mDeepFRI for orthogonal function prediction. All accept FASTA. Scientifically novel but niche. PEbA's pairwise limitation makes it better as a benchmark tool than production search.

### E. Uni-Mol Active Learning for Lead Optimization
Uni-Mol Repr as features for workflow-009's surrogate model. 2026 paper validates this combination. More of an internal upgrade to workflow-009.

---

## What to Do

1. **Pick a direction** (or propose a different combination)
2. **Sketch a node structure** using the proposal format from issue #90
3. **Post as a comment on this issue**

I'll review and we'll finalize the scope together.

---

## References
- **workflow-025 (Vina vs GNINA):** best reference for a parallel comparison + report workflow
- **workflow-014 (ADMET-AI):** reference for SMILES-based property prediction pipeline
- **workflow-012 (Boltz-2):** reference for structure prediction workflow
- **Silva Migration Guide:** `SILVA_MIGRATION_GUIDE.md` in repo root

**Comment by allisongcheng** [comment]
## Structure Prediction → Pocket Prediction → Docking
### Boltz-2 → P2Rank → Uni-Mol Docking V2 (when no crystal structure exists) 
- GitHub: https://github.com/deepmodeling/Uni-Mol
- Install: `pip install unimol_tools`
- What it does: A universal 3D molecular representation framework that utilizes SE(3)-equivariant neural networks for high-accuracy protein-ligand docking, 3D embedding extraction, and molecular property prediction. It excels at geometric deep learning tasks, outperforming traditional physics-based baselines by directly learning from 3D molecular conformations.

### Proposed Node Structure
1. `01_boltz2_predict` — Generates the 3D protein structure coordinates and per-residue confidence scores (pLDDT) directly from a raw amino acid sequence using deep learning.
2. `02_p2rank_pocket_find` — Analyzes the predicted protein structure's surface topology using machine learning to identify and locate coordinates of potential ligand-binding pockets.
3. `03_pocket_qc_grid` — Cross-references the predicted pocket coordinates with Boltz-2's pLDDT metrics, filters out unstable or low-confidence pocket regions (pLDDT < 70), and maps the optimal centroid into a standardized bounding box JSON configuration.
4. `04_unimol_docking` — Takes the structural target and ligand input to execute geometry-driven, deep learning-accelerated docking within the validated pocket coordinates, outputting finalized 3D binding poses.
5. `05_visualize` — Renders an interactive 3D molecular visualization of the predicted protein fold, highlighting the P2Rank-identified pocket colored by its pLDDT confidence, alongside the top-scoring Uni-Mol ligand conformations.

### Input / Output
- Input: Protein FASTA sequence (.fasta) + Ligand structural file (.sdf or .smiles)
- Output: Filtered, high-confidence docked protein-ligand complex poses (.sdf) + Docking affinity/confidence score report (.json)

### Why this tool?
Integrating Uni-Mol Docking V2 with Boltz-2 and P2Rank creates an entirely autonomous, AI-driven structure-to-docking pipeline. Relying solely on predicted structures introduces a major challenge: traditional docking engines often fail or give false positives when forced into pockets that fall within poorly modeled, flexible, or intrinsically disordered regions. By building an intelligent QC node that maps geometric pocket coordinates directly back to the spatial confidence (pLDDT) of the structural model, this workflow ensures that downstream deep-learning docking is only performed on highly reliable binding pockets.

**Comment by RongjunG** [comment]
Hi @allisongcheng, thank you for the detailed proposal! Based on it, I did some research into the official docs and recent papers for each tool, sharing what I found, along with a few things to adjust.

---

### Input Format

Boltz-2 requires **YAML input** (not FASTA) for protein-ligand complex prediction. FASTA mode doesn't support ligand specification or affinity prediction. This is consistent with how workflow-012 handles Boltz-2 input.

Example:
```yaml
version: 1
sequences:
  - protein:
      id: A
      sequence: MVLSPADKTNVKA...
  - ligand:
      id: B
      smiles: 'CC(=O)Oc1ccccc1'
properties:
  - affinity:
      binder: B
```

Also worth noting: recent literature (AlphaFold3 VS paper, biorxiv 2025) shows that **holo prediction** (with ligand present in the input) significantly improves binding site quality over apo prediction. So including the ligand SMILES in the YAML is not just for affinity — it helps structure prediction too.

**Suggested input**: `protein.yaml` (containing sequence + ligand SMILES + affinity block) instead of `.fasta` + `.sdf/.smiles` separately.

---

### I/O Compatibility & Missing Steps

I checked the exact I/O formats for each tool. The chain mostly works, but there are a few gaps in the current proposal:

Following the node order (Boltz-2 → P2Rank → QC → Uni-Mol):

**Node 01 → Node 02 (Boltz-2 → P2Rank): ✅ Direct compatibility**
- Boltz-2 outputs mmCIF (`.cif`). Note: the `--output_format pdb` flag is currently broken (GitHub issue [#298](https://github.com/jwohlwend/boltz/issues/298), IndexError on protein-ligand complexes). Always use mmCIF output.
- P2Rank natively accepts mmCIF — no conversion needed.
- P2Rank should use the **`-c alphafold` profile** since the input comes from Boltz-2 (B-factor column contains pLDDT, not thermal displacement).

**Node 02 → Node 03 (P2Rank → Pocket QC): ⚠️ Needs box size calculation**
- P2Rank outputs `*_predictions.csv` with `center_x, center_y, center_z` (SAS point centroid in Å) — these map directly to Uni-Mol's grid JSON `center_x/y/z`.
- However, P2Rank does **not** output box size. Uni-Mol's grid JSON requires `size_x, size_y, size_z` as well. Node 03 needs to calculate these from the pocket's spatial extent (typical approach: pocket SAS radius + 5–10 Å padding → 15–25 Å box).

**Node 03 → Node 04 (QC → Uni-Mol): ⚠️ Format conversions needed in node 03**
- **mmCIF → PDB**: Uni-Mol Docking V2 requires PDB input (does not support mmCIF). `gemmi convert input.cif output.pdb` is the simplest option. Avoid BioPython's `MMCIFIO` for writing back Boltz-2 CIF (known compatibility issue with missing `entity.id`).
- **SMILES → SDF**: Uni-Mol requires SDF for the ligand. If the input is SMILES (from the Boltz-2 YAML), node 03 should handle SMILES → SDF conversion (RDKit `AllChem.EmbedMolecule`).
- Known issue: Uni-Mol's RDKit sanitization can **silently drop invalid ligands** ([issue #281](https://github.com/deepmodeling/Uni-Mol/issues/281)). Adding a validation check before passing to Uni-Mol would be safer.

---

### Uni-Mol Docking V2 Output — No Affinity Score

One important finding: **Uni-Mol Docking V2 outputs only 3D poses (SDF), with no binding affinity or confidence score.** The internal `prmsd_score` used for pose ranking is not written to the output file.

This means the "Docking affinity/confidence score report (.json)" described in the proposal can't come directly from Uni-Mol. Instead, we can rely on **Boltz-2's own confidence metrics**, which are quite rich:

- `plddt_*.npz` — per-residue structural confidence (0–1)
- `pae_*.npz` — predicted aligned error matrix
- `confidence_*.json` — aggregate complex confidence
- `affinity_*.json` — predicted binding affinity (if the YAML includes an `affinity` block)

Combined with P2Rank's pocket `score` and `probability`, there's enough for a meaningful report without adding a separate scoring node.

Ref:
- [Uni-Mol Docking V2 README](https://github.com/deepmodeling/Uni-Mol/tree/main/unimol_docking_v2)
- Boltz-2 confidence output: `plddt`, `pae`, `pde`, `affinity` files per model

---

### pLDDT ≥ 70 Threshold

The threshold is well-supported:
- AlphaFold's official scale defines ≥70 as "confident, backbone correct"
- PrankWeb 4 (Jakubec et al. 2025, *NAR*) uses pLDDT > 70 as its default filtering toggle
- Nature Communications 2024 druggable pocket study: pocket centroid 8 Å residues all pLDDT > 70

**One caveat worth noting** — Eguida & Rognan 2023 (*JCIM*, [PMC9852548](https://pmc.ncbi.nlm.nih.gov/articles/PMC9852548/)) tested 22 drug targets and found that 4 of the 5 worst-performing targets for docking had binding-site pLDDT ≥ 70. High pLDDT is necessary but not sufficient — it filters disordered regions but doesn't catch subtle side-chain misalignment or loop occlusion.

Suggestion: in node 03, report the **per-pocket pLDDT distribution** (mean, min, std across pocket residues) rather than just a binary pass/fail, so the user can make an informed judgment. The report should include a note that pLDDT ≥ 70 is a standard but imperfect filter.

Also, for reading pLDDT: use `plddt_*.npz` (per-token, 0–1 scale), not `confidence_*.json` (that's an aggregate value, not per-residue).

---

### Report Node (05)

For the report/visualization node, since we've used Molstar in other workflows, I'd suggest sticking with that for the 3D viewer. Combined with Plotly + Bootstrap (similar to workflow-014's pattern), the report could include:

1. **Pipeline summary** — input sequence length, model count, pocket count, QC-passed pockets, docked compounds
2. **Structure confidence panel** — Boltz-2 pLDDT distribution histogram + PAE heatmap (Plotly)
3. **Pocket discovery table** — P2Rank rank, score, probability, mean pLDDT per pocket, center coords (sortable)
4. **3D interactive viewer** — Molstar: protein colored by pLDDT, pocket highlighted, top docking pose overlaid
5. **Methods & caveats** — auto-generated description + "pLDDT ≥ 70 is necessary but not sufficient" note

Ref for report patterns:
- dockviz (ljmartin, 2024) — PDBe-Molstar embedded in HTML for docking visualization
- PrankWeb 4 — pocket surface colored by P2Rank score with pLDDT toggle
- workflow-014 (ADMET-AI) — Plotly radar charts + Bootstrap 5 layout pattern

---

### Containerization Notes

- **Uni-Core** (Uni-Mol's dependency) requires source compilation with strict CUDA version matching. Base image: `dptechnology/unicore:0.0.1-pytorch1.11.0-cuda11.3`. Model weights (464 MB) need manual download from Dropbox — no pip/HuggingFace distribution.
- **P2Rank** needs Java runtime in the container.
- **Boltz-2**: existing workflow-012 image (`ghcr.io/chiral-data/boltz:2025_09_05`) can be reused for node 01.

---

### Example Input

One more thing — it would be good to include a concrete example input with the proposal. From my research, I'd suggest picking a target that **has an existing crystal structure with a co-crystallized ligand** (holo structure in PDB). The reason: even though this pipeline is designed for cases where no crystal structure exists, having a known experimental structure lets us validate the output — we can check whether Boltz-2's predicted structure matches the crystal, whether P2Rank finds the correct pocket, and whether Uni-Mol's docking pose is close to the real binding mode.

Could you look into finding a suitable target? Ideally something with a well-resolved holo structure (≤2.5 Å) and a drug-like ligand with known SMILES.

---

### Summary of Suggested Changes

| # | Item | Change |
|---|---|---|
| 1 | Input format | `.fasta` → `.yaml` (with ligand SMILES + affinity block) |
| 2 | Node 03 | Add mmCIF → PDB conversion (`gemmi convert`) |
| 3 | Node 03 | Add grid box size calculation (P2Rank only outputs center, not size) |
| 4 | Node 03 | Add SMILES → SDF conversion if ligand input is SMILES |
| 5 | Node 02 | Use P2Rank `-c alphafold` profile |
| 6 | Output spec | Remove "docking affinity score" from Uni-Mol output — use Boltz-2 metrics instead |
| 7 | Node 05 | Expand from "visualize" to full report (Molstar + Plotly + summary tables + caveats) |

Overall the workflow proposal is approved! Feel free to refer to above before start building it to make the implementation a bit smoother.
